arithmetization/src/main/java/net/consensys/linea/zktracer/module/shakiradata/ShakiraDataOperation.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright ConsenSys Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package net.consensys.linea.zktracer.module.shakiradata;
+
+import static net.consensys.linea.zktracer.types.Conversions.bigIntegerToBytes;
+import static net.consensys.linea.zktracer.types.Conversions.bytesToUnsignedBytes;
+import static net.consensys.linea.zktracer.types.Utils.leftPadTo;
+import static net.consensys.linea.zktracer.types.Utils.rightPadTo;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import com.google.common.base.Preconditions;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+import net.consensys.linea.zktracer.container.ModuleOperation;
+import net.consensys.linea.zktracer.types.UnsignedByte;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.crypto.Hash;
+
+@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
+@Accessors(fluent = true)
+public class ShakiraDataOperation extends ModuleOperation {
+  private static final int PHASE_EXTRA = 0;
+  private static final int INDEX_MAX_EXTRA = 2;
+  private static final int TOTAL_SIZE_RESULT_PHASES = 32;
+  private static final int TOTAL_SIZE_EXTRA_PHASES = 0;
+  public static final int RESULT_ROWS_COUNT = 2;
+  public static final int EXTRA_ROWS_COUNT = 3;
+
+  @Getter @EqualsAndHashCode.Include private final int hubStamp;
+  @Getter @Setter private int prevId;
+  @Getter @Setter private int currentId;
+
+  @EqualsAndHashCode.Include private final ShakiraPrecompileType precompileType;
+  @EqualsAndHashCode.Include private final Bytes hashInput;
+
+  public ShakiraDataOperation(int hubStamp, ShakiraPrecompileType precompileType, Bytes hashInput) {
+    this.hubStamp = hubStamp;
+    this.precompileType = precompileType;
+    this.hashInput = hashInput;
+  }
+
+  @Override
+  protected int computeLineCount() {
+    return totalNumberOfRows();
+  }
+
+  void trace(Trace trace, int stamp) {
+    var tracerBuilder = ShakiraTraceHelper.builder().trace(trace).currentId(currentId).stamp(stamp);
+
+    switch (precompileType) {
```

Comment:
I don't understand why we duplicate for each precompile, we trace exactly the same for the three operation (only the flag and the way we compute the result differ)